packages/webapp/src/_app/eos/secret-publisher.ts

```diff
@@ -0,0 +1,240 @@
+        encryptedSessionKeys,
+        publisherKey,
+        recipientKeys,
+        transientKeyPair,
+        encryptedMessage,
+    });
+
+    // TODO: remove this test
+    console.info("testing decryption...");
+    await decryptPublishedMessage(
+        encryptedMessage,
+        publisherKey,
+        transientKeyPair.publicKey.toString(),
+        encryptedSessionKeys[0],
+        info
+    );
+
+    return {};
+};
+
+export const decryptPublishedMessage = async (
+    encryptedMessage: Uint8Array,
+    recipientPublicKey: string,
+    transientPublicKey: string,
+    encryptedSessionKey: Uint8Array,
+    info?: string
+) => {
+    const recipientPrivateKey = retrieveRecipientPrivateKey(recipientPublicKey);
+    const ecdhSecret = deriveEcdhSecret(
+        recipientPrivateKey,
+        PublicKey.fromString(transientPublicKey)
+    );
+    console.info("ecdh secret is ", ecdhSecret);
+    const hkdfKey = await hkdfSha256FromEcdh(ecdhSecret, info);
+    console.info("hkdf key to unwrap is", hkdfKey);
+    const sessionKey = await unwrapSessionKey(encryptedSessionKey, hkdfKey);
+    console.info("unwrapped session key!", sessionKey);
+    const message = await decryptMessage(sessionKey, encryptedMessage);
+    console.info("decrypted message:", message);
+    return message;
+};
+
+const retrieveRecipientPrivateKey = (publicKey: string): PrivateKey => {
+    // TODO: find the corresponding private key in the localstorage, or else throw
+    console.info("retrieving publicKey...");
+    const rawPrivateKey = "5KcMypBPGByXbeBphsVzyyzUvg31tWNKsgrfXs2MmanYgYf4gao";
+    return PrivateKey.fromString(rawPrivateKey);
+};
+
+const fetchRecipientKeys = async (accounts: string[]) => {
+    // TODO: real implementation
+    // fetch eden.keys_table( where account = each account item )
+
+    const mockedKeysLen = accounts.length;
+    const mockedKeys = [];
+    for (let i = 0; i < mockedKeysLen; i++) {
+        mockedKeys.push(
+            "EOS5wGY8RcG8PuGYTPU9QwmiEm9XReBkK4SGtSuS4f1oeX4TNGaHT"
+        );
+        // temp PK: 5KcMypBPGByXbeBphsVzyyzUvg31tWNKsgrfXs2MmanYgYf4gao
+    }
+    return mockedKeys;
+};
+
+const ecdhRecipientsKeyEncriptionKeys = async (
+    recipientPublicKeys: string[],
+    transientPrivateKey: PrivateKey,
+    info?: string
+): Promise<CryptoKey[]> => {
+    return Promise.all(
+        recipientPublicKeys.map(async (publicKey: string) => {
+            const ecdhSecret = deriveEcdhSecret(
+                transientPrivateKey,
+                PublicKey.fromString(publicKey)
+            );
+            console.info(ecdhSecret);
+            return await hkdfSha256FromEcdh(ecdhSecret, info);
+        })
+    );
+};
+
+const deriveEcdhSecret = (
+    eosPrivateKeyA: PrivateKey,
+    eosPublicKeyB: PublicKey
+) => {
+    return eosPrivateKeyA
+        .toElliptic()
+        .derive(eosPublicKeyB.toElliptic().getPublic())
+        .toArrayLike(ArrayBuffer); // TODO: review endianness, empty = Big Endian
```

Comment:
Since this is treated as pseudo-random data by HKDF, the endianness doesn't matter

Reply:
If we want to allow third-party front ends to interact with this data (or if we later decide to re-write this using a different API), then it matters, because every implementation needs to use the same endianness.